### PR TITLE
TK, DSSM, PACRR rerankers + MSMarco collection

### PR DESCRIPTION
**Pull request by kevinmartinjos**
Adds the following rerankers:
 - TK. Results are [here](https://docs.google.com/spreadsheets/d/1xkah1HjpCdsJvNPZUGHVvKx5Mqnqri8DP6vaavD_GEU/edit#gid=0)
 - PACRR, DSSM. See the rows highlighted in red [here](https://docs.google.com/spreadsheets/d/1WlGjZmgi22QipEwkWlCSFOl2BDQnkpyHNTLu2mt5vhM/edit#gid=0)

Adds MSMarco dataset. Pending: Run tk on MSMarco. Takes way too long :(

Adds additional caching logic (`extractor.usecache`) for EmbedText and BagOfWords extractor. 

Adds a `validatefreq` config flag to the trainer so that validation does not (optionally) run after every iteration. Adds a `fastforward` config flag to disable fast forwards

Removes the git lfs command from the github workflow (actions) and downloads the qrels/folds from [dropbox](https://www.dropbox.com/s/h6iovqwsyyv7660/data.tgz?dl=0).

Ran `black` on the entire project.

### CODE REVIEW

File: capreolus/benchmark/__init__.py [hunk 1]
```diff
@@ -63,3 +63,11 @@ class ANTIQUE(Benchmark):
     topic_file = PACKAGE_PATH / "data" / "topics.antique.txt"
     fold_file = PACKAGE_PATH / "data" / "antique.json"
     query_type = "title"
+
+
+class MSMarcoPassage(Benchmark):
```
Comment (kevinmartinjos) [inline capreolus/benchmark/__init__.py]: Msmarco

File: capreolus/extractor/__init__.py [hunk 2]
```diff
@@ -37,13 +48,46 @@ def _extend_stoi(self, toks_list, calc_idf=False):
...
+    def load_state(self, qids, docids):
+        raise NotImplementedError
+
+    def get_state_cache_file_path(self, qids, docids):
+        """
+        Returns the path to the cache file used to store the extractor state, regardless of whether it exists or not
+        """
+        sorted_qids = sorted(qids)
+        sorted_docids = sorted(docids)
+        return (
```
Comment (kevinmartinjos) [inline capreolus/extractor/__init__.py]: New caching logic

File: capreolus/trainer/__init__.py [hunk 3]
```diff
@@ -21,23 +21,29 @@ class Trainer(ModuleBase, metaclass=RegisterableModule):
...
-        interactive = False  # True for training with Notebook or False for command line environment
-
+        softmaxloss = (
+            False
+        )  # True to use softmax loss (over pairs) or False to use hinge loss
+
+        interactive = (
+            False
+        )  # True for training with Notebook or False for command line environment
+        fastforward = False
```
Comment (kevinmartinjos) [inline capreolus/trainer/__init__.py]: The new config flags

File: .github/workflows/pythonpackage.yml [hunk 4]
```diff
@@ -21,6 +21,14 @@ jobs:
       uses: actions/setup-java@v1
       with:
               java-version: '11'
+    - name: pull .txt files
+      run: |
+        cd capreolus/data
+        rm -rf dummy
+        rm ./*
+        wget https://www.dropbox.com/s/h6iovqwsyyv7660/data.tgz?dl=1
```
Comment (kevinmartinjos) [inline .github/workflows/pythonpackage.yml]: No longer uses git lfs. The caveat is that we should remember to update the link if folds/qrels ever change or if we add a new dataset. This is not really that much of a big deal since the unit tests only make use of the dummy dataset and the dummy dataset would never really change.

File: capreolus/extractor/bagofwords.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    """ Bag of Words (or bag of trigrams when `datamode=trigram`) extractor. Used with the DSSM reranker. """
+
+    name = "bagofwords"
+    dependencies = {
+        "index": Dependency(module="index", name="anserini", config_overrides={"indexstops": True, "stemmer": "none"}),
+        "tokenizer": Dependency(module="tokenizer", name="anserini"),
+    }
+    pad = 0
+    pad_tok = "<pad>"
+    tokenizer_name = "anserini"
```
Comment (andrewyates) [inline capreolus/extractor/bagofwords.py]: Is this used anywhere?

File: capreolus/extractor/bagofwords.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    @staticmethod
+    def config():
+        datamode = "unigram"  # type of input: 'unigram' or 'trigram'
+        keepstops = False  # include stopwords in the reranker's input
```
Comment (andrewyates) [inline capreolus/extractor/bagofwords.py]: I think this isn't used -- it was before the rewrite, but now it's a config option of the tokenizer dependency

File: capreolus/registry.py [hunk 7]
```diff
@@ -183,6 +183,8 @@ def get_cache_path(self):
     def get_module_path(self, include_provided=True):
         """ Return a path encoding the module's config, including its dependenceis """
 
+        # TODO: Potential bug - self.moudles is an ordinary dict and hence keys are not ordered. Can result in different
+        # cache paths
```
Comment (andrewyates) [inline capreolus/registry.py]: Good catch. Will fix in a separate PR cleaning up the pipeline init

File: capreolus/trainer/__init__.py [hunk 8]
```diff
@@ -225,20 +228,22 @@ def train(self, reranker, train_dataset, train_output_path, dev_data, dev_output
...
-
+            iters_so_far += 1
             # predict performance on dev set
-            pred_fn = dev_output_path / f"{niter}.run"
-            preds = self.predict(reranker, dev_data, pred_fn)
 
-            # log dev metrics
-            metrics = evaluator.eval_runs(preds, qrels, ["ndcg_cut_20", "map", "P_20"])
-            logger.info("dev metrics: %s", " ".join([f"{metric}={v:0.3f}" for metric, v in sorted(metrics.items())]))
+            if iters_so_far % validation_frequency == 0:
```
Comment (andrewyates) [inline capreolus/trainer/__init__.py]: Maybe this should be based on `niters` rather than `iters_so_far`, so that the same iters are checked against the validation set even if `fastforward=True` causes training to be resumed?

File: capreolus/extractor/bagofwords.py (hunk 5 above)
Comment (kevinmartinjos) [reply capreolus/extractor/bagofwords.py]: No I was copy pasting things from the old code.
